core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -2823,73 +2914,313 @@ public void cleanUp() {
+                    LOGGER.log(SEVERE, "Failed " + t.getDisplayName(), err);
+                }
+
+                public void onAttained(Milestone milestone) {
+                    Level lv = level;
+                    String s = "Attained " + milestone.toString();
+                    if (milestone instanceof TermMilestone) {
+                        lv = Level.INFO; // noteworthy milestones --- at least while we debug problems further
+                        s = milestone.toString();
+                    }
+                    LOGGER.log(lv, s);
+                }
             });
-        } catch (InterruptedException e) {
-            LOGGER.log(SEVERE, "Failed to execute termination",e);
-            e.printStackTrace();
-        } catch (ReactorException e) {
-            LOGGER.log(SEVERE, "Failed to execute termination",e);
-        } catch (IOException e) {
+        } catch (InterruptedException | ReactorException | IOException e) {
             LOGGER.log(SEVERE, "Failed to execute termination",e);
+            errors.add(e);
+        } catch (OutOfMemoryError e) {
+            // we should just propagate this, no point trying to log
+            throw e;
+        } catch (LinkageError e) {
+            LOGGER.log(SEVERE, "Failed to execute termination", e);
+            // safe to ignore and continue for this one
+        } catch (Throwable e) {
+            LOGGER.log(SEVERE, "Failed to execute termination", e);
+            // save for later
+            errors.add(e);
         }
+    }
 
+    private Set<Future<?>> _cleanUpDisconnectComputers(final List<Throwable> errors) {
+        LOGGER.log(Level.INFO, "Starting node disconnection");
         final Set<Future<?>> pending = new HashSet<Future<?>>();
-        terminating = true;
         // JENKINS-28840 we know we will be interrupting all the Computers so get the Queue lock once for all
         Queue.withLock(new Runnable() {
             @Override
             public void run() {
                 for( Computer c : computers.values() ) {
-                    c.interrupt();
-                    killComputer(c);
-                    pending.add(c.disconnect(null));
+                    try {
+                        c.interrupt();
+                        killComputer(c);
+                        pending.add(c.disconnect(null));
+                    } catch (OutOfMemoryError e) {
+                        // we should just propagate this, no point trying to log
+                        throw e;
+                    } catch (LinkageError e) {
+                        LOGGER.log(Level.WARNING, "Could not disconnect " + c + ": " + e.getMessage(), e);
+                        // safe to ignore and continue for this one
+                    } catch (Throwable e) {
+                        LOGGER.log(Level.WARNING, "Could not disconnect " + c + ": " + e.getMessage(), e);
+                        // save for later
+                        errors.add(e);
```

Comment:
:ant: Yikes, this is a lot of duplication. Why not introduce a helper method taking `Callable<Void>` (lambdas, eventually) and handling the boilerplate of catching and reporting exceptions?